swift/ql/lib/codeql/swift/printast/PrintAstNode.qll

```diff
@@ -74,13 +80,86 @@ class PrintLocatable extends PrintAstNode, TLocatable {
 
   PrintLocatable() { this = TLocatable(ast) }
 
+  override string toString() { result = prettyPrint(ast) }
+
   final override predicate shouldBePrinted() { shouldPrint(ast) }
 
   override predicate hasChild(PrintAstNode child, int index, string accessor) {
     child = TLocatable(getChildAndAccessor(ast, index, accessor))
   }
 
-  override string toString() { result = "[" + concat(ast.getPrimaryQlClasses(), ", ") + "] " + ast }
-
   final override Location getLocation() { result = ast.getLocation() }
 }
+
+/**
+ * A graph node representing a conversion.
+ */
+class PrintConversion extends PrintAstNode, TConversion {
+  Expr conv;
+
+  PrintConversion() { this = TConversion(conv) }
+
+  override string toString() { result = prettyPrint(conv) }
+
+  final override predicate shouldBePrinted() { shouldPrint(conv.resolve()) }
+
+  override predicate hasChild(PrintAstNode child, int index, string accessor) { none() }
+
+  final override Location getLocation() { result = conv.getLocation() }
+}
+
+/**
+ * A graph node representing a virtual container for conversions.
+ */
+class PrintConversionContainer extends PrintAstNode, TConversionContainer {
+  Expr convertee;
+
+  PrintConversionContainer() { this = TConversionContainer(convertee) }
+
+  override string toString() { result = "" }
+
+  final override predicate shouldBePrinted() { shouldPrint(convertee) }
+
+  override predicate hasChild(PrintAstNode child, int index, string accessor) {
+    child = TConversion(convertee.getConversion(index)) and
+    accessor = "getConversion(" + index + ")"
```

Comment:
I think there's a usability question to be resolved here.
The `accessor` string is meant to inform users about which predicate they should be calling to get the corresponding AST element, and I'm not sure we want to advocate calls to `getConversion(1)`, `getConversion(2)`, ... `getConversion(n)`.

Instead, I think we should have a `getFullyConverted()` child under which the entire AST is laid out (including hidden AST nodes and whatever). IIRC, this is what we did for C/C++.

Do you think that could work for Swift? And would it be better?

Reply:
hmm, as far as I understand the PrintAst on C++ only shows one conversion, which I found quite confusing when multiple convertions are present.

For more similarity to the C++ case, we could maybe go with a repeated `Conversion` label, starting with the `getFullyConverted` one, like (I wonder how will that work, and also if there is a `semmle.order` thing to do on properties as well to get that)
```
#  155|       getInit(0): [CallExpr] call to init()
#  155|           Conversion: [InjectIntoOptionalExpr] (Baz?) ...
#  155|           Conversion: [CoerceExpr] (Baz) ...
#  155|           Conversion: [DerivedToBaseExpr] (Baz) ...
#  155|         getFunction(): [ConstructorRefCallExpr] call to init()
#  155|           getFunction(): [DeclRefExpr] init()
#  155|           getArgument(0): [Argument] : Derived.Type
#  155|             getExpr(): [TypeExpr] Derived.Type
#  155|               getTypeRepr(): [TypeRepr] Derived
```
When there is one conversion this would look exactly like C++

Reply:
That's an incorrect observation regarding C++. You'll get nested conversions:
```
# 1685|           getExpr().getFullyConverted(): [ReferenceToExpr] (reference to)
# 1685|               Type = [LValueReferenceType] const CapturedLambdaMyObj &
# 1685|               ValueCategory = prvalue
# 1685|             getExpr(): [CStyleCast] (const CapturedLambdaMyObj)...
# 1685|                 Conversion = [GlvalueConversion] glvalue conversion
# 1685|                 Type = [SpecifiedType] const CapturedLambdaMyObj
# 1685|                 ValueCategory = lvalue
# 1685|               getExpr(): [TemporaryObjectExpr] temporary object
# 1685|                   Type = [Class] CapturedLambdaMyObj
# 1685|                   ValueCategory = lvalue
```

Reply:
Yeah, I think I like the C/C++ visualization better since it's clearer which predicate should be called to get to the conversion of interest through `getFullyConverted`.

With that said, if you think this is difficult to model in the current Swift setup then I'm perfectly happy with leaving that reordering as future work as this PR is clearly better than no PrintAST support for conversions 😂.

Reply:
ah, ok I copmletely misunderstood the C++ format then. I can try reproduce that.

Reply:
it's done. If we fix the fact that contrary to C++ the getters in the hidden AST also resolve nodes in Swift, we can get rid of the `Immediate`, but I'd do it in another PR

Reply:
Sounds good to me!